drivers/usb/usbuhci_new/usbuhci.c

```diff
@@ -0,0 +1,2791 @@
+                                              FALSE,
+                                              &LegacySupport.AsUSHORT,
+                                              PCI_LEGSUP,
+                                              sizeof(USHORT));
+    }
+
+    return MpStatus;
+}
+
+MPSTATUS
+NTAPI
+UhciInitializeHardware(IN PUHCI_EXTENSION UhciExtension)
+{
+    PUHCI_HW_REGISTERS BaseRegister;
+    UHCI_USB_COMMAND Command;
+    UHCI_USB_STATUS StatusMask;
+
+    DPRINT("UhciInitializeHardware: UhciExtension - %p\n", UhciExtension);
+    DPRINT("UhciInitializeHardware: VIA HW FIXME\n"); // after supporting HcFlavor in usbport
+
+    BaseRegister = UhciExtension->BaseRegister;
+
+    /* Save SOF Timing Value */
+    UhciExtension->SOF_Modify = READ_PORT_UCHAR(&BaseRegister->SOF_Modify);
+
+    RegPacket.UsbPortWait(UhciExtension, 20);
+
+    Command.AsUSHORT = READ_PORT_USHORT(&BaseRegister->HcCommand.AsUSHORT);
+
+    /* Global Reset */
+    Command.AsUSHORT = 0;
+    Command.GlobalReset = 1;
+    WRITE_PORT_USHORT(&BaseRegister->HcCommand.AsUSHORT, Command.AsUSHORT);
+
+    RegPacket.UsbPortWait(UhciExtension, 20);
+
+    Command.AsUSHORT = 0;
+    WRITE_PORT_USHORT(&BaseRegister->HcCommand.AsUSHORT, Command.AsUSHORT);
+
+    /* Set MaxPacket for full speed bandwidth reclamation */
+    Command.AsUSHORT = 0;
+    Command.MaxPacket = 1; // 64 bytes
+    WRITE_PORT_USHORT(&BaseRegister->HcCommand.AsUSHORT, Command.AsUSHORT);
+
+    /* Restore SOF Timing Value */
+    WRITE_PORT_UCHAR(&BaseRegister->SOF_Modify, UhciExtension->SOF_Modify);
+
+    StatusMask = UhciExtension->StatusMask;
+
+    StatusMask.Interrupt = 1;
+    StatusMask.ErrorInterrupt = 1;
+    StatusMask.ResumeDetect = 1;
+    StatusMask.HostSystemError = 1;
+
+    UhciExtension->StatusMask = StatusMask;
+
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+UhciInitializeSchedule(IN PUHCI_EXTENSION UhciExtension,
+                       IN PUHCI_HC_RESOURCES HcResourcesVA,
+                       IN ULONG hcResourcesPA)
+{
+    PUHCI_HCD_QH IntQH;
+    ULONG IntQhPA;
+    PUHCI_HCD_QH StaticControlHead;
+    ULONG StaticControlHeadPA;
+    PUHCI_HCD_QH StaticBulkHead;
+    ULONG StaticBulkHeadPA;
+    PUHCI_HCD_TD StaticBulkTD;
+    ULONG StaticBulkTdPA;
+    PUHCI_HCD_TD StaticTD;
+    ULONG StaticTdPA;
+    PUHCI_HCD_TD StaticSofTD;
+    ULONG StaticSofTdPA;
+    ULONG PhysicalAddress;
+    ULONG Idx;
+    ULONG HeadIdx;
+    UCHAR FrameIdx;
+    PUHCI_HC_RESOURCES HcResourcesPA = (PUHCI_HC_RESOURCES)hcResourcesPA;
+
+    DPRINT("UhciInitializeSchedule: Ext[%p], VA - %p, PA - %p\n",
+           UhciExtension,
+           HcResourcesVA,
+           HcResourcesPA);
+
+    /* Build structure (tree) of static QHs
+       for interrupt and isochronous transfers */
+    for (FrameIdx = 0; FrameIdx < INTERRUPT_ENDPOINTs; FrameIdx++)
+    {
+        IntQH = &HcResourcesVA->StaticIntHead[FrameIdx];
+        IntQhPA = (ULONG)&HcResourcesPA->StaticIntHead[FrameIdx];
```

Comment:
I see a lot of pointy things being cast to ULONG. Couldnt review deeply atm.

Reply:
USB controllers use 32 bit DMA.